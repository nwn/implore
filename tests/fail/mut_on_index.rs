use implore::impl_op;

#[derive(Copy, Clone, PartialEq, Eq, Debug)]
struct X;

#[impl_op]
fn index(x: &X, y: u32) -> X {
    X
}

#[impl_op]
fn index(x: X, y: u32) -> &X {
    &mut X
}

#[impl_op]
fn index(x: &mut X, y: u32) -> &X {
    &X
}

#[impl_op]
fn index(x: &X, y: u32) -> &mut X {
    &mut X
}

fn main() {}
