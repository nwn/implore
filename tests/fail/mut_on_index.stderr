error: `index` must return a reference type
 --> $DIR/mut_on_index.rs:7:28
  |
7 | fn index(x: &X, y: u32) -> X {
  |                            ^

error: the first operand of `index` must be a reference
  --> $DIR/mut_on_index.rs:12:13
   |
12 | fn index(x: X, y: u32) -> &X {
   |             ^

error: the first operand of `index` must be an immutable reference
  --> $DIR/mut_on_index.rs:17:13
   |
17 | fn index(x: &mut X, y: u32) -> &X {
   |             ^^^^^^

error: `index` must return an immutable reference type
  --> $DIR/mut_on_index.rs:22:28
   |
22 | fn index(x: &X, y: u32) -> &mut X {
   |                            ^^^^^^
